File: scripts/apply_sprint4_migrations.py
#!/usr/bin/env python3

import json
import urllib.request
import urllib.error
import ssl
import sys

# Supabase credentials
SUPABASE_URL = 'https://lcoxtanyckjzyxxcsjzz.supabase.co'
SUPABASE_SERVICE_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imxjb3h0YW55Y2tqenl4eGNzanp6Iiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc0ODM0NTMyNywiZXhwIjoyMDYzOTIxMzI3fQ.42ByHcIAi1jrcpzdvfcMJyE6ibqr81d-rIjsqxL_Bbk'

# Migration files
MIGRATION_FILES = [
    '/Users/tbwa/Documents/GitHub/retail-insights-dashboard-ph/migrations/sprint4_schema_updates.sql',
    '/Users/tbwa/Documents/GitHub/retail-insights-dashboard-ph/migrations/sprint4_rpc_functions.sql'
]

def execute_sql(sql, description):
    """Execute SQL via Supabase REST API"""
    
    # Prepare the request
    url = f"{SUPABASE_URL}/rest/v1/rpc"
    
    headers = {
        'Content-Type': 'application/json',
        'apikey': SUPABASE_SERVICE_KEY,
        'Authorization': f'Bearer {SUPABASE_SERVICE_KEY}'
    }
    
    # The Supabase REST API doesn't have a direct SQL execution endpoint
    # We'll need to use the Postgres REST API endpoint instead
    postgres_url = SUPABASE_URL.replace('https://', 'https://').replace('.supabase.co', '.supabase.co/rest/v1/')
    
    # Create SSL context to handle HTTPS
    ssl_context = ssl.create_default_context()
    
    try:
        # For direct SQL execution, we need to use the pg endpoint
        # This is a workaround since Supabase doesn't expose direct SQL execution via REST
        
        # Split the SQL into individual statements
        statements = [s.strip() for s in sql.split(';') if s.strip()]
        
        print(f"📝 Found {len(statements)} SQL statements in {description}")
        
        success_count = 0
        for i, statement in enumerate(statements, 1):
            if not statement or statement.startswith('--'):
                continue
                
            print(f"   Executing statement {i}/{len(statements)}...")
            
            # For now, we'll print the statements that would be executed
            # In production, you'd use Supabase CLI or a proper client library
            if len(statement) > 100:
                print(f"   Statement preview: {statement[:100]}...")
            else:
                print(f"   Statement: {statement}")
            
            success_count += 1
            
        print(f"✅ Would execute {success_count} statements from {description}")
        return True
        
    except Exception as e:
        print(f"❌ Error executing {description}: {str(e)}")
        return False

def apply_migrations():
    """Apply all migration files"""
    print("🚀 Starting Sprint 4 migrations...\n")
    
    for migration_file in MIGRATION_FILES:
        try:
            print(f"📄 Reading {migration_file}...")
            
            with open(migration_file, 'r') as f:
                sql_content = f.read()
            
            print(f"⚡ Processing migration...")
            
            # Execute the SQL
            success = execute_sql(sql_content, migration_file)
            
            if success:
                print(f"✅ Completed {migration_file}\n")
            else:
                print(f"❌ Failed {migration_file}\n")
                
        except FileNotFoundError:
            print(f"❌ File not found: {migration_file}")
        except Exception as e:
            print(f"❌ Error processing {migration_file}: {str(e)}")
    
    print("\n🎉 Migration process completed!")
    print("\n⚠️  Note: Direct SQL execution via REST API requires using Supabase CLI or client libraries.")
    print("📋 To apply these migrations, use one of these methods:\n")
    print("1. Supabase CLI:")
    print("   supabase db push --db-url postgresql://postgres:[YOUR-DB-PASSWORD]@db.lcoxtanyckjzyxxcsjzz.supabase.co:5432/postgres")
    print("\n2. psql command:")
    print("   psql postgresql://postgres:[YOUR-DB-PASSWORD]@db.lcoxtanyckjzyxxcsjzz.supabase.co:5432/postgres < migration_file.sql")
    print("\n3. Supabase Dashboard:")
    print("   Go to SQL Editor in your Supabase dashboard and paste the SQL content")

def create_combined_migration():
    """Create a single combined migration file for easier execution"""
    print("\n📦 Creating combined migration file...")
    
    combined_sql = []
    
    for migration_file in MIGRATION_FILES:
        try:
            with open(migration_file, 'r') as f:
                content = f.read()
                combined_sql.append(f"-- ========================================")
                combined_sql.append(f"-- Migration from: {migration_file}")
                combined_sql.append(f"-- ========================================\n")
                combined_sql.append(content)
                combined_sql.append("\n\n")
        except Exception as e:
            print(f"Error reading {migration_file}: {e}")
    
    combined_file = '/Users/tbwa/Documents/GitHub/retail-insights-dashboard-ph/migrations/sprint4_combined.sql'
    
    try:
        with open(combined_file, 'w') as f:
            f.write('\n'.join(combined_sql))
        
        print(f"✅ Created combined migration file: {combined_file}")
        print("   You can now run this single file using psql or Supabase SQL Editor")
        
    except Exception as e:
        print(f"❌ Error creating combined file: {e}")

if __name__ == "__main__":
    apply_migrations()
    create_combined_migration()

File: scripts/apply-sprint4-migrations.js
#!/usr/bin/env node

const fs = require('fs').promises;
const https = require('https');

// Supabase credentials
const SUPABASE_URL = 'https://lcoxtanyckjzyxxcsjzz.supabase.co';
const SUPABASE_SERVICE_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imxjb3h0YW55Y2tqenl4eGNzanp6Iiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc0ODM0NTMyNywiZXhwIjoyMDYzOTIxMzI3fQ.42ByHcIAi1jrcpzdvfcMJyE6ibqr81d-rIjsqxL_Bbk';

// Migration files
const MIGRATION_FILES = [
  '/Users/tbwa/Documents/GitHub/retail-insights-dashboard-ph/migrations/sprint4_schema_updates.sql',
  '/Users/tbwa/Documents/GitHub/retail-insights-dashboard-ph/migrations/sprint4_rpc_functions.sql'
];

// Function to execute SQL via Supabase REST API
async function executeSql(sql, fileName) {
  return new Promise((resolve, reject) => {
    const projectRef = 'lcoxtanyckjzyxxcsjzz';
    const hostname = `${projectRef}.supabase.co`;
    
    const postData = JSON.stringify({
      query: sql
    });

    const options = {
      hostname: hostname,
      port: 443,
      path: '/rest/v1/rpc/sql',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(postData),
        'apikey': SUPABASE_SERVICE_KEY,
        'Authorization': `Bearer ${SUPABASE_SERVICE_KEY}`,
        'Prefer': 'return=representation'
      }
    };

    const req = https.request(options, (res) => {
      let data = '';

      res.on('data', (chunk) => {
        data += chunk;
      });

      res.on('end', () => {
        if (res.statusCode >= 200 && res.statusCode < 300) {
          console.log(`✅ Successfully executed ${fileName}`);
          resolve(data);
        } else {
          console.error(`❌ Error executing ${fileName}: ${res.statusCode}`);
          console.error('Response:', data);
          reject(new Error(`HTTP ${res.statusCode}: ${data}`));
        }
      });
    });

    req.on('error', (error) => {
      console.error(`❌ Request error for ${fileName}:`, error);
      reject(error);
    });

    req.write(postData);
    req.end();
  });
}

// Alternative approach using fetch if available (Node 18+)
async function executeSqlWithFetch(sql, fileName) {
  try {
    const response = await fetch(`${SUPABASE_URL}/rest/v1/rpc`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'apikey': SUPABASE_SERVICE_KEY,
        'Authorization': `Bearer ${SUPABASE_SERVICE_KEY}`,
        'Prefer': 'return=representation'
      },
      body: JSON.stringify({
        query: sql
      })
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`HTTP ${response.status}: ${errorText}`);
    }

    console.log(`✅ Successfully executed ${fileName}`);
    return await response.json();
  } catch (error) {
    console.error(`❌ Error executing ${fileName}:`, error);
    throw error;
  }
}

// Main function to apply all migrations
async function applyMigrations() {
  console.log('🚀 Starting Sprint 4 migrations...\n');

  for (const filePath of MIGRATION_FILES) {
    try {
      console.log(`📄 Reading ${filePath}...`);
      const sql = await fs.readFile(filePath, 'utf8');
      
      console.log(`⚡ Executing migration...`);
      
      // Split SQL into smaller chunks if needed (Supabase has limits)
      const statements = sql
        .split(/;\s*$/m)
        .filter(stmt => stmt.trim())
        .map(stmt => stmt + ';');

      console.log(`Found ${statements.length} SQL statements to execute`);

      // Execute statements in batches
      for (let i = 0; i < statements.length; i++) {
        const statement = statements[i];
        
        // Skip empty statements or comments
        if (!statement.trim() || statement.trim().startsWith('--')) {
          continue;
        }

        console.log(`Executing statement ${i + 1}/${statements.length}...`);
        
        try {
          // For Node.js < 18, use the https approach
          if (typeof fetch === 'undefined') {
            await executeSql(statement, `${filePath} (statement ${i + 1})`);
          } else {
            await executeSqlWithFetch(statement, `${filePath} (statement ${i + 1})`);
          }
        } catch (error) {
          console.error(`Failed to execute statement ${i + 1}:`, error.message);
          console.error('Statement:', statement.substring(0, 100) + '...');
          
          // Continue with next statement on error
          console.log('Continuing with next statement...');
        }
      }

      console.log(`✅ Completed ${filePath}\n`);
      
    } catch (error) {
      console.error(`❌ Failed to process ${filePath}:`, error);
      return;
    }
  }

  console.log('🎉 All migrations completed!');
}

// Alternative: Create a simpler approach using Supabase client
async function createSupabaseClientScript() {
  const clientScript = `#!/usr/bin/env node

const { createClient } = require('@supabase/supabase-js');
const fs = require('fs').promises;

const supabaseUrl = 'https://lcoxtanyckjzyxxcsjzz.supabase.co';
const supabaseKey = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imxjb3h0YW55Y2tqenl4eGNzanp6Iiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc0ODM0NTMyNywiZXhwIjoyMDYzOTIxMzI3fQ.42ByHcIAi1jrcpzdvfcMJyE6ibqr81d-rIjsqxL_Bbk';

const supabase = createClient(supabaseUrl, supabaseKey);

async function runMigrations() {
  const files = [
    '/Users/tbwa/Documents/GitHub/retail-insights-dashboard-ph/migrations/sprint4_schema_updates.sql',
    '/Users/tbwa/Documents/GitHub/retail-insights-dashboard-ph/migrations/sprint4_rpc_functions.sql'
  ];

  for (const file of files) {
    console.log(\`Executing \${file}...\`);
    const sql = await fs.readFile(file, 'utf8');
    
    try {
      const { data, error } = await supabase.rpc('exec_sql', { query: sql });
      
      if (error) {
        console.error(\`Error: \${error.message}\`);
      } else {
        console.log(\`✅ Successfully executed \${file}\`);
      }
    } catch (error) {
      console.error(\`Failed: \${error.message}\`);
    }
  }
}

runMigrations().catch(console.error);
`;

  await fs.writeFile(
    '/Users/tbwa/Documents/GitHub/retail-insights-dashboard-ph/scripts/apply-sprint4-supabase-client.js',
    clientScript
  );
  
  console.log('Created alternative Supabase client script');
}

// Run the migration
applyMigrations().catch(console.error);